// @lucid-agents/hono/dist/index.js
import { Hono } from 'hono';
import { paymentMiddlewareFromConfig } from '@x402/hono';
import { ExactEvmScheme } from '@x402/evm/exact/server';
import { HTTPFacilitatorClient } from '@x402/core/server';
import { resolvePrice, validatePaymentsConfig, resolvePayTo, extractSenderDomain, evaluateSender, extractPayerAddress, parsePriceAmount, findMostSpecificIncomingLimit, createFacilitatorAuthHeaders } from '@lucid-agents/payments';

// src/app.ts
var DEFAULT_SCHEME = "exact";
var DEFAULT_SCHEMES = [
  {
    network: "eip155:*",
    server: new ExactEvmScheme()
  }
];
function withPaymentHeader(c, paymentHeader) {
  const mergedHeaders = new Headers(c.req.raw.headers);
  if (!mergedHeaders.has("X-PAYMENT")) {
    mergedHeaders.set("X-PAYMENT", paymentHeader);
  }
  const requestWithPayment = new Request(c.req.raw, {
    headers: mergedHeaders
  });
  const originalHeader = c.req.header.bind(c.req);
  const reqProxy = new Proxy(c.req, {
    get(target, prop, receiver) {
      if (prop === "raw") {
        return requestWithPayment;
      }
      if (prop === "header") {
        return (name) => {
          if (!name) return originalHeader();
          if (name.toLowerCase() === "x-payment") {
            return requestWithPayment.headers.get("X-PAYMENT") ?? void 0;
          }
          return originalHeader(name);
        };
      }
      return Reflect.get(target, prop, receiver);
    }
  });
  const contextProxy = new Proxy(c, {
    get(target, prop, receiver) {
      if (prop === "req") {
        return reqProxy;
      }
      return Reflect.get(target, prop, receiver);
    }
  });
  return contextProxy;
}
function addFacilitatorAuth(facilitator, token) {
  if (facilitator.createAuthHeaders) {
    return facilitator;
  }
  const authHeaders = createFacilitatorAuthHeaders(token);
  if (!authHeaders) {
    return facilitator;
  }
  return {
    ...facilitator,
    createAuthHeaders: async () => authHeaders
  };
}
function withPayments({
  app,
  path,
  entrypoint,
  kind,
  payments,
  facilitator,
  middlewareFactory = paymentMiddlewareFromConfig,
  runtime
}) {
  if (!payments) return false;
  const network = entrypoint.network ?? payments.network;
  const price = resolvePrice(entrypoint, payments, kind);
  validatePaymentsConfig(payments, network, entrypoint.key);
  if (!price) return false;
  const payTo = resolvePayTo(payments);
  const description = entrypoint.description ?? `${entrypoint.key}${kind === "stream" ? " (stream)" : ""}`;
  const postMimeType = kind === "stream" ? "text/event-stream" : "application/json";
  const baseFacilitator = facilitator ?? { url: payments.facilitatorUrl };
  const resolvedFacilitator = addFacilitatorAuth(
    baseFacilitator,
    payments.facilitatorAuth
  );
  const postRoute = {
    accepts: {
      scheme: DEFAULT_SCHEME,
      payTo,
      price,
      network
    },
    description,
    mimeType: postMimeType
  };
  const getRoute = {
    accepts: {
      scheme: DEFAULT_SCHEME,
      payTo,
      price,
      network
    },
    description,
    mimeType: "application/json"
  };
  const policyGroups = runtime?.payments?.policyGroups;
  const paymentTracker = runtime?.payments?.paymentTracker;
  if (policyGroups && policyGroups.length > 0) {
    app.use(path, async (c, next) => {
      const senderDomain = extractSenderDomain(
        c.req.header("origin"),
        c.req.header("referer")
      );
      for (const group of policyGroups) {
        if (group.blockedSenders || group.allowedSenders) {
          const result = evaluateSender(group, void 0, senderDomain);
          if (!result.allowed) {
            return c.json(
              {
                error: {
                  code: "policy_violation",
                  message: result.reason || "Payment blocked by policy",
                  groupName: result.groupName
                }
              },
              403
            );
          }
        }
      }
      await next();
    });
  }
  const facilitatorClient = new HTTPFacilitatorClient(resolvedFacilitator);
  const routes = {
    [`POST ${path}`]: postRoute,
    [`GET ${path}`]: getRoute
  };
  const baseMiddleware = middlewareFactory(
    routes,
    facilitatorClient,
    DEFAULT_SCHEMES
  );
  app.use(path, async (c, next) => {
    const paymentHeader = c.req.header("PAYMENT");
    const contextForPayment = paymentHeader && !c.req.header("X-PAYMENT") ? withPaymentHeader(c, paymentHeader) : c;
    const result = await baseMiddleware(contextForPayment, next);
    if (result instanceof Response) {
      return result;
    }
  });
  if (policyGroups && policyGroups.length > 0 && paymentTracker) {
    app.use(path, async (c, next) => {
      await next();
      const paymentResponseHeader = c.res.headers.get("PAYMENT-RESPONSE");
      if (paymentResponseHeader && c.res.status >= 200 && c.res.status < 300) {
        try {
          const payerAddress = extractPayerAddress(paymentResponseHeader);
          const senderDomain = extractSenderDomain(
            c.req.header("origin"),
            c.req.header("referer")
          );
          const paymentAmount = parsePriceAmount(price);
          if (payerAddress && paymentAmount !== void 0) {
            for (const group of policyGroups) {
              if (group.incomingLimits) {
                const limitInfo = findMostSpecificIncomingLimit(
                  group.incomingLimits,
                  payerAddress,
                  senderDomain,
                  c.req.url
                );
                const scope = limitInfo?.scope ?? "global";
                await paymentTracker.recordIncoming(
                  group.name,
                  scope,
                  paymentAmount
                );
              }
            }
          }
        } catch (error) {
          console.error("[paywall] Error recording incoming payment:", error);
        }
      }
    });
  }
  return true;
}

// src/app.ts
async function createAgentApp(runtime, opts) {
  if (!runtime.handlers) {
    throw new Error(
      "HTTP extension is required. Use app.use(http()) when building the runtime."
    );
  }
  const app = new Hono();
  opts?.beforeMount?.(app);
  const registerEntrypointRoutes = (entrypoint) => {
    const invokePath = `/entrypoints/${entrypoint.key}/invoke`;
    const streamPath = `/entrypoints/${entrypoint.key}/stream`;
    withPayments({
      app,
      path: invokePath,
      entrypoint,
      kind: "invoke",
      payments: runtime.payments?.config,
      runtime
    });
    app.post(
      invokePath,
      (c) => runtime.handlers.invoke(c.req.raw, { key: entrypoint.key })
    );
    withPayments({
      app,
      path: streamPath,
      entrypoint,
      kind: "stream",
      payments: runtime.payments?.config,
      runtime
    });
    app.post(
      streamPath,
      (c) => runtime.handlers.stream(c.req.raw, { key: entrypoint.key })
    );
  };
  app.get("/health", (c) => runtime.handlers.health(c.req.raw));
  app.get("/entrypoints", (c) => runtime.handlers.entrypoints(c.req.raw));
  app.get("/.well-known/agent.json", (c) => runtime.handlers.manifest(c.req.raw));
  app.get(
    "/.well-known/agent-card.json",
    (c) => runtime.handlers.manifest(c.req.raw)
  );
  app.get(
    "/.well-known/oasf-record.json",
    (c) => runtime.handlers.oasf ? runtime.handlers.oasf(c.req.raw) : c.json(
      {
        error: {
          code: "not_found",
          message: "OASF record is not enabled"
        }
      },
      404
    )
  );
  app.get("/favicon.svg", (c) => runtime.handlers.favicon(c.req.raw));
  app.post("/tasks", (c) => runtime.handlers.tasks(c.req.raw));
  app.get("/tasks", (c) => runtime.handlers.listTasks(c.req.raw));
  app.get(
    "/tasks/:taskId",
    (c) => runtime.handlers.getTask(c.req.raw, { taskId: c.req.param("taskId") })
  );
  app.post(
    "/tasks/:taskId/cancel",
    (c) => runtime.handlers.cancelTask(c.req.raw, { taskId: c.req.param("taskId") })
  );
  app.get(
    "/tasks/:taskId/subscribe",
    (c) => runtime.handlers.subscribeTask(c.req.raw, { taskId: c.req.param("taskId") })
  );
  if (runtime.handlers.landing) {
    app.get("/", (c) => runtime.handlers.landing(c.req.raw));
  } else {
    app.get("/", (c) => c.text("Landing disabled", 404));
  }
  const addEntrypoint = (def) => {
    runtime.entrypoints.add(def);
    const entrypoint = runtime.entrypoints.snapshot().find((item) => item.key === def.key);
    if (!entrypoint) {
      throw new Error(`Failed to register entrypoint "${def.key}"`);
    }
    registerEntrypointRoutes(entrypoint);
  };
  for (const entrypoint of runtime.entrypoints.snapshot()) {
    registerEntrypointRoutes(entrypoint);
  }
  opts?.afterMount?.(app);
  const result = {
    app,
    runtime,
    agent: runtime.agent,
    addEntrypoint
  };
  return result;
}

export { createAgentApp, withPayments };
